File: Ejercicio3.py
import tkinter as tk
from tkinter import messagebox, ttk

class Categoria:
    def __init__(self, nombre):
        self.nombre = nombre

    def mostrar_info(self):
        return f"Categoría: {self.nombre}"

class Producto:
    def __init__(self, nombre, precio, categoria):
        self.nombre = nombre
        self.precio = precio
        self.categoria = categoria

    def mostrar_info(self):
        return f"Producto: {self.nombre}, Precio: {self.precio}, {self.categoria.mostrar_info()}"

class Cliente:
    def __init__(self, nombre, apellido, id_cliente):
        self.nombre = nombre
        self.apellido = apellido
        self.id_cliente = id_cliente

    def mostrar_info(self):
        return f"Cliente: {self.nombre} {self.apellido}, ID: {self.id_cliente}"

class ItemOrden:
    def __init__(self, producto, cantidad):
        self.producto = producto
        self.cantidad = cantidad

    def calcular_subtotal(self):
        return self.producto.precio * self.cantidad

class Orden:
    def __init__(self, cliente):
        self.cliente = cliente
        self.items = []
        self.total = 0

    def agregar_item(self, item):
        self.items.append(item)
        self.calcular_total()

    def calcular_total(self):
        self.total = sum(item.calcular_subtotal() for item in self.items)

class Tienda:
    def __init__(self):
        self.productos = []
        self.clientes = []
        self.ordenes = []
        self.categorias = []

    def registrar_producto(self, producto):
        self.productos.append(producto)

    def registrar_cliente(self, cliente):
        self.clientes.append(cliente)

    def crear_orden(self, orden):
        self.ordenes.append(orden)

    def mostrar_productos(self):
        return [producto.mostrar_info() for producto in self.productos]

    def mostrar_clientes(self):
        return [cliente.mostrar_info() for cliente in self.clientes]

    def mostrar_ordenes(self):
        ordenes_info = []
        for orden in self.ordenes:
            items_info = [f"{item.producto.nombre} x{item.cantidad}" for item in orden.items]
            ordenes_info.append(f"Orden de {orden.cliente.nombre} {orden.cliente.apellido}: {', '.join(items_info)}, Total: {orden.total}")
        return ordenes_info

class TiendaGUI:
    def __init__(self, root):
        self.tienda = Tienda()
        self.root = root
        self.root.title("Sistema de Gestión de Tienda")
        self.root.configure(bg="#f0f0f0")
        self.create_main_menu()

    def create_main_menu(self):
        self.clear_frame()
        frame = ttk.Frame(self.root, padding="10 10 10 10")
        frame.place(relx=0.5, rely=0.5, anchor="center")

        ttk.Button(frame, text="Registrar Producto", command=self.registrar_producto, width=30).pack(pady=10)
        ttk.Button(frame, text="Registrar Cliente", command=self.registrar_cliente, width=30).pack(pady=10)
        ttk.Button(frame, text="Crear Orden", command=self.crear_orden, width=30).pack(pady=10)
        ttk.Button(frame, text="Mostrar Productos", command=self.mostrar_productos, width=30).pack(pady=10)
        ttk.Button(frame, text="Mostrar Clientes", command=self.mostrar_clientes, width=30).pack(pady=10)
        ttk.Button(frame, text="Mostrar Ordenes", command=self.mostrar_ordenes, width=30).pack(pady=10)

    def registrar_producto(self):
        self.clear_frame()
        frame = ttk.Frame(self.root, padding="10 10 10 10")
        frame.place(relx=0.5, rely=0.5, anchor="center")

        ttk.Label(frame, text="Registrar Producto", font=("Helvetica", 16)).pack(pady=10)
        ttk.Label(frame, text="Nombre").pack(pady=5)
        nombre_entry = ttk.Entry(frame, width=50)
        nombre_entry.pack(pady=5)

        ttk.Label(frame, text="Precio").pack(pady=5)
        precio_entry = ttk.Entry(frame, width=50)
        precio_entry.pack(pady=5)

        ttk.Label(frame, text="Categoría").pack(pady=5)
        categoria_entry = ttk.Entry(frame, width=50)
        categoria_entry.pack(pady=5)

        def submit():
            nombre = nombre_entry.get()
            precio = float(precio_entry.get())
            nombre_categoria = categoria_entry.get()

            categoria = next((cat for cat in self.tienda.categorias if cat.nombre == nombre_categoria), None)
            if not categoria:
                categoria = Categoria(nombre_categoria)
                self.tienda.categorias.append(categoria)

            producto = Producto(nombre, precio, categoria)
            self.tienda.registrar_producto(producto)
            messagebox.showinfo("Éxito", "Producto registrado correctamente.")
            self.create_main_menu()

        ttk.Button(frame, text="Registrar", command=submit).pack(pady=20)
        ttk.Button(frame, text="Volver", command=self.create_main_menu).pack()

    def registrar_cliente(self):
        self.clear_frame()
        frame = ttk.Frame(self.root, padding="10 10 10 10")
        frame.place(relx=0.5, rely=0.5, anchor="center")

        ttk.Label(frame, text="Registrar Cliente", font=("Helvetica", 16)).pack(pady=10)
        ttk.Label(frame, text="Nombre").pack(pady=5)
        nombre_entry = ttk.Entry(frame, width=50)
        nombre_entry.pack(pady=5)

        ttk.Label(frame, text="Apellido").pack(pady=5)
        apellido_entry = ttk.Entry(frame, width=50)
        apellido_entry.pack(pady=5)

        ttk.Label(frame, text="ID Cliente").pack(pady=5)
        id_cliente_entry = ttk.Entry(frame, width=50)
        id_cliente_entry.pack(pady=5)

        def submit():
            nombre = nombre_entry.get()
            apellido = apellido_entry.get()
            id_cliente = id_cliente_entry.get()

            cliente = Cliente(nombre, apellido, id_cliente)
            self.tienda.registrar_cliente(cliente)
            messagebox.showinfo("Éxito", "Cliente registrado correctamente.")
            self.create_main_menu()

        ttk.Button(frame, text="Registrar", command=submit).pack(pady=20)
        ttk.Button(frame, text="Volver", command=self.create_main_menu).pack()

    def crear_orden(self):
        self.clear_frame()
        frame = ttk.Frame(self.root, padding="10 10 10 10")
        frame.place(relx=0.5, rely=0.5, anchor="center")

        ttk.Label(frame, text="Crear Orden", font=("Helvetica", 16)).pack(pady=10)
        ttk.Label(frame, text="ID Cliente").pack(pady=5)
        id_cliente_entry = ttk.Entry(frame, width=50)
        id_cliente_entry.pack(pady=5)

        items_frame = ttk.Frame(frame)
        items_frame.pack(pady=5)
        items = []

        def agregar_item():
            producto_nombre = producto_entry.get()
            cantidad = int(cantidad_entry.get())

            producto = next((prod for prod in self.tienda.productos if prod.nombre == producto_nombre), None)
            if producto:
                items.append(ItemOrden(producto, cantidad))
                item_label = ttk.Label(items_frame, text=f"{producto_nombre} x{cantidad}")
                item_label.pack()

        ttk.Label(frame, text="Producto").pack(pady=5)
        producto_entry = ttk.Entry(frame, width=50)
        producto_entry.pack(pady=5)

        ttk.Label(frame, text="Cantidad").pack(pady=5)
        cantidad_entry = ttk.Entry(frame, width=50)
        cantidad_entry.pack(pady=5)

        ttk.Button(frame, text="Agregar Item", command=agregar_item).pack(pady=10)

        def submit():
            id_cliente = id_cliente_entry.get()
            cliente = next((cli for cli in self.tienda.clientes if cli.id_cliente == id_cliente), None)
            if cliente:
                orden = Orden(cliente)
                for item in items:
                    orden.agregar_item(item)
                self.tienda.crear_orden(orden)
                messagebox.showinfo("Éxito", "Orden creada correctamente.")
            else:
                messagebox.showerror("Error", "Cliente no encontrado.")
            self.create_main_menu()

        ttk.Button(frame, text="Crear Orden", command=submit).pack(pady=20)
        ttk.Button(frame, text="Volver", command=self.create_main_menu).pack()

    def mostrar_productos(self):
        self.clear_frame()
        frame = ttk.Frame(self.root, padding="10 10 10 10")
        frame.place(relx=0.5, rely=0.5, anchor="center")

        ttk.Label(frame, text="Productos Registrados", font=("Helvetica", 16)).pack(pady=10)
        productos = self.tienda.mostrar_productos()
        for producto in productos:
            ttk.Label(frame, text=producto).pack(pady=5)

        ttk.Button(frame, text="Volver", command=self.create_main_menu).pack(pady=20)

    def mostrar_clientes(self):
        self.clear_frame()
        frame = ttk.Frame(self.root, padding="10 10 10 10")
        frame.place(relx=0.5, rely=0.5, anchor="center")

        ttk.Label(frame, text="Clientes Registrados", font=("Helvetica", 16)).pack(pady=10)
        clientes = self.tienda.mostrar_clientes()
        for cliente in clientes:
            ttk.Label(frame, text=cliente).pack(pady=5)

        ttk.Button(frame, text="Volver", command=self.create_main_menu).pack(pady=20)

    def mostrar_ordenes(self):
        self.clear_frame()
        frame = ttk.Frame(self.root, padding="10 10 10 10")
        frame.place(relx=0.5, rely=0.5, anchor="center")

        ttk.Label(frame, text="Ordenes Registradas", font=("Helvetica", 16)).pack(pady=10)
        ordenes = self.tienda.mostrar_ordenes()
        for orden in ordenes:
            ttk.Label(frame, text=orden).pack(pady=5)

        ttk.Button(frame, text="Volver", command=self.create_main_menu).pack(pady=20)

    def clear_frame(self):
        for widget in self.root.winfo_children():
            widget.destroy()

if __name__ == "__main__":
    root = tk.Tk()
    app = TiendaGUI(root)
    root.mainloop()

File: Ejercicio3/src/ejercicio3/Ejercicio3.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Main.java to edit this template
 */
package ejercicio3;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.util.ArrayList;

public class Ejercicio3 {

    static class Categoria {
        String nombre;

        Categoria(String nombre) {
            this.nombre = nombre;
        }

        String mostrarInfo() {
            return "Categoría: " + nombre;
        }
    }

    static class Producto {
        String nombre;
        double precio;
        Categoria categoria;

        Producto(String nombre, double precio, Categoria categoria) {
            this.nombre = nombre;
            this.precio = precio;
            this.categoria = categoria;
        }

        String mostrarInfo() {
            return "Producto: " + nombre + ", Precio: " + precio + ", " + categoria.mostrarInfo();
        }
    }

    static class Cliente {
        String nombre;
        String apellido;
        String idCliente;

        Cliente(String nombre, String apellido, String idCliente) {
            this.nombre = nombre;
            this.apellido = apellido;
            this.idCliente = idCliente;
        }

        String mostrarInfo() {
            return "Cliente: " + nombre + " " + apellido + ", ID: " + idCliente;
        }
    }

    static class ItemOrden {
        Producto producto;
        int cantidad;

        ItemOrden(Producto producto, int cantidad) {
            this.producto = producto;
            this.cantidad = cantidad;
        }

        double calcularSubtotal() {
            return producto.precio * cantidad;
        }
    }

    static class Orden {
        Cliente cliente;
        ArrayList<ItemOrden> items;
        double total;

        Orden(Cliente cliente) {
            this.cliente = cliente;
            this.items = new ArrayList<>();
            this.total = 0;
        }

        void agregarItem(ItemOrden item) {
            items.add(item);
            calcularTotal();
        }

        void calcularTotal() {
            total = 0;
            for (ItemOrden item : items) {
                total += item.calcularSubtotal();
            }
        }
    }

    static class Tienda {
        ArrayList<Producto> productos;
        ArrayList<Cliente> clientes;
        ArrayList<Orden> ordenes;
        ArrayList<Categoria> categorias;

        Tienda() {
            productos = new ArrayList<>();
            clientes = new ArrayList<>();
            ordenes = new ArrayList<>();
            categorias = new ArrayList<>();
        }

        void registrarProducto(Producto producto) {
            productos.add(producto);
        }

        void registrarCliente(Cliente cliente) {
            clientes.add(cliente);
        }

        void crearOrden(Orden orden) {
            ordenes.add(orden);
        }

        ArrayList<String> mostrarProductos() {
            ArrayList<String> productosInfo = new ArrayList<>();
            for (Producto producto : productos) {
                productosInfo.add(producto.mostrarInfo());
            }
            return productosInfo;
        }

        ArrayList<String> mostrarClientes() {
            ArrayList<String> clientesInfo = new ArrayList<>();
            for (Cliente cliente : clientes) {
                clientesInfo.add(cliente.mostrarInfo());
            }
            return clientesInfo;
        }

        ArrayList<String> mostrarOrdenes() {
            ArrayList<String> ordenesInfo = new ArrayList<>();
            for (Orden orden : ordenes) {
                StringBuilder itemsInfo = new StringBuilder();
                for (ItemOrden item : orden.items) {
                    itemsInfo.append(item.producto.nombre).append(" x").append(item.cantidad).append(", ");
                }
                ordenesInfo.add("Orden de " + orden.cliente.nombre + " " + orden.cliente.apellido + ": " +
                        itemsInfo.toString() + "Total: " + orden.total);
            }
            return ordenesInfo;
        }
    }

    static class TiendaGUI {
        JFrame frame;
        Tienda tienda;

        TiendaGUI() {
            tienda = new Tienda();
            frame = new JFrame("Sistema de Gestión de Tienda");
            frame.setSize(600, 400);
            frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
            frame.setLayout(new BorderLayout());

            createMainMenu();
        }

        void createMainMenu() {
            JPanel panel = new JPanel(new GridLayout(6, 1));
            panel.setBorder(BorderFactory.createEmptyBorder(10, 10, 10, 10));

            JButton registrarProductoBtn = new JButton("Registrar Producto");
            registrarProductoBtn.addActionListener(e -> registrarProducto());
            panel.add(registrarProductoBtn);

            JButton registrarClienteBtn = new JButton("Registrar Cliente");
            registrarClienteBtn.addActionListener(e -> registrarCliente());
            panel.add(registrarClienteBtn);

            JButton crearOrdenBtn = new JButton("Crear Orden");
            crearOrdenBtn.addActionListener(e -> crearOrden());
            panel.add(crearOrdenBtn);

            JButton mostrarProductosBtn = new JButton("Mostrar Productos");
            mostrarProductosBtn.addActionListener(e -> mostrarProductos());
            panel.add(mostrarProductosBtn);

            JButton mostrarClientesBtn = new JButton("Mostrar Clientes");
            mostrarClientesBtn.addActionListener(e -> mostrarClientes());
            panel.add(mostrarClientesBtn);

            JButton mostrarOrdenesBtn = new JButton("Mostrar Ordenes");
            mostrarOrdenesBtn.addActionListener(e -> mostrarOrdenes());
            panel.add(mostrarOrdenesBtn);

            frame.add(panel, BorderLayout.CENTER);
            frame.setVisible(true);
        }

        void registrarProducto() {
            JTextField nombreField = new JTextField(5);
            JTextField precioField = new JTextField(5);
            JComboBox<String> categoriaBox = new JComboBox<>();
            for (Categoria categoria : tienda.categorias) {
                categoriaBox.addItem(categoria.nombre);
            }

            JPanel myPanel = new JPanel();
            myPanel.add(new JLabel("Nombre:"));
            myPanel.add(nombreField);
            myPanel.add(Box.createHorizontalStrut(15)); // a spacer
            myPanel.add(new JLabel("Precio:"));
            myPanel.add(precioField);
            myPanel.add(new JLabel("Categoría:"));
            myPanel.add(categoriaBox);

            int result = JOptionPane.showConfirmDialog(null, myPanel,
                    "Registrar Producto", JOptionPane.OK_CANCEL_OPTION);
            if (result == JOptionPane.OK_OPTION) {
                String nombre = nombreField.getText();
                double precio = Double.parseDouble(precioField.getText());
                String categoriaNombre = (String) categoriaBox.getSelectedItem();
                Categoria categoria = tienda.categorias.stream()
                        .filter(c -> c.nombre.equals(categoriaNombre))
                        .findFirst()
                        .orElse(null);
                if (categoria != null) {
                    Producto producto = new Producto(nombre, precio, categoria);
                    tienda.registrarProducto(producto);
                    JOptionPane.showMessageDialog(frame, "Producto registrado con éxito");
                } else {
                    JOptionPane.showMessageDialog(frame, "Categoría no encontrada");
                }
            }
        }

        void registrarCliente() {
            JTextField nombreField = new JTextField(5);
            JTextField apellidoField = new JTextField(5);
            JTextField idClienteField = new JTextField(5);

            JPanel myPanel = new JPanel();
            myPanel.add(new JLabel("Nombre:"));
            myPanel.add(nombreField);
            myPanel.add(Box.createHorizontalStrut(15)); // a spacer
            myPanel.add(new JLabel("Apellido:"));
            myPanel.add(apellidoField);
            myPanel.add(new JLabel("ID Cliente:"));
            myPanel.add(idClienteField);

            int result = JOptionPane.showConfirmDialog(null, myPanel,
                    "Registrar Cliente", JOptionPane.OK_CANCEL_OPTION);
            if (result == JOptionPane.OK_OPTION) {
                String nombre = nombreField.getText();
                String apellido = apellidoField.getText();
                String idCliente = idClienteField.getText();
                Cliente cliente = new Cliente(nombre, apellido, idCliente);
                tienda.registrarCliente(cliente);
                JOptionPane.showMessageDialog(frame, "Cliente registrado con éxito");
            }
        }

        void crearOrden() {
            JComboBox<String> clienteBox = new JComboBox<>();
            for (Cliente cliente : tienda.clientes) {
                clienteBox.addItem(cliente.nombre + " " + cliente.apellido);
            }

            JComboBox<String> productoBox = new JComboBox<>();
            for (Producto producto : tienda.productos) {
                productoBox.addItem(producto.nombre);
            }

            JTextField cantidadField = new JTextField(5);

            JPanel myPanel = new JPanel();
            myPanel.add(new JLabel("Cliente:"));
            myPanel.add(clienteBox);
            myPanel.add(new JLabel("Producto:"));
            myPanel.add(productoBox);
            myPanel.add(Box.createHorizontalStrut(15)); // a spacer
            myPanel.add(new JLabel("Cantidad:"));
            myPanel.add(cantidadField);

            int result = JOptionPane.showConfirmDialog(null, myPanel,
                    "Crear Orden", JOptionPane.OK_CANCEL_OPTION);
            if (result == JOptionPane.OK_OPTION) {
                String clienteNombre = (String) clienteBox.getSelectedItem();
                String[] nombreApellido = clienteNombre.split(" ");
                Cliente cliente = tienda.clientes.stream()
                        .filter(c -> c.nombre.equals(nombreApellido[0]) && c.apellido.equals(nombreApellido[1]))
                        .findFirst()
                        .orElse(null);

                String productoNombre = (String) productoBox.getSelectedItem();
                Producto producto = tienda.productos.stream()
                        .filter(p -> p.nombre.equals(productoNombre))
                        .findFirst()
                        .orElse(null);

                int cantidad = Integer.parseInt(cantidadField.getText());

                if (cliente != null && producto != null) {
                    Orden orden = new Orden(cliente);
                    ItemOrden itemOrden = new ItemOrden(producto, cantidad);
                    orden.agregarItem(itemOrden);
                    tienda.crearOrden(orden);
                    JOptionPane.showMessageDialog(frame, "Orden creada con éxito");
                } else {
                    JOptionPane.showMessageDialog(frame, "Cliente o Producto no encontrados");
                }
            }
        }

        void mostrarProductos() {
            ArrayList<String> productosInfo = tienda.mostrarProductos();
            JOptionPane.showMessageDialog(frame, String.join("\n", productosInfo), "Productos", JOptionPane.INFORMATION_MESSAGE);
        }

        void mostrarClientes() {
            ArrayList<String> clientesInfo = tienda.mostrarClientes();
            JOptionPane.showMessageDialog(frame, String.join("\n", clientesInfo), "Clientes", JOptionPane.INFORMATION_MESSAGE);
        }

        void mostrarOrdenes() {
            ArrayList<String> ordenesInfo = tienda.mostrarOrdenes();
            JOptionPane.showMessageDialog(frame, String.join("\n", ordenesInfo), "Órdenes", JOptionPane.INFORMATION_MESSAGE);
        }
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(TiendaGUI::new);
    }
}
